File: CSharp/Thing.cs
using System;
using System.Collections.Generic;
using System.Linq;
using ThingModel.Builders;

namespace ThingModel
{
    public class Thing
    {
        /**
         * String ID, unique in the world.
         * 
         * Is never null.
         */
        private readonly string _id;
        public string ID { get { return _id; } }

        /**
         * Reference to the type of the object.
         * 
         * Can be null if the thing is related to the default type.
         */
        private readonly ThingType _type;
        public ThingType Type { get { return _type; } }

        /**
         * A thing contains a list of properties.
         * A property is defined by a key and a value.
         */
        protected IDictionary<string, Property> Properties { get; private set; }

        /**
         * A thing can be connected to other things,
         * but never directly to itself.
         */
        protected IDictionary<string, Thing> Connections { get; private set; } 

        /**
         * Create a new thing. The id cannot be null and should
         * be unique in the world. If the type is null, a default
         * thing type is used.
         */
        public Thing(string id, ThingType type = null)
        {
            if (string.IsNullOrEmpty(id))
            {
                throw new Exception("The ID should not be null or empty");
            }

            _id = id;
            _type = type;
            Properties = new Dictionary<string, Property>();
            Connections = new Dictionary<string, Thing>();
        }

        public void SetProperty(Property property)
        {
            Properties[property.Key] = property;
        }

        public T GetProperty<T>(string key) where T : Property
        {
            Property value;
            Properties.TryGetValue(key, out value);
            return value as T;
        }

        public bool HasProperty(string key)
        {
            return Properties.ContainsKey(key);
        }

        public void Connect(Thing thing)
        {
	        if (thing == null) return;

            if (thing == this || thing._id == _id)
            {
                throw new Exception("Cannot connect a thing to itself");
            }
            Connections.Add(thing._id, thing);
        }

        public bool Disconnect(Thing thing)
        {
	        return thing != null && Connections.Remove(thing._id);
        }

	    public bool IsConnectedTo(Thing thing)
        {
	        if (thing == null) return false;
            return Connections.ContainsKey(thing._id);
        }

        public void Detach()
        {
            Connections.Clear();
        }

        /**
         * Get a list of connected things.
         * The list is a copy so you can connect and disconnect
         * things when you are enumating the list.
         */
        public IList<Thing> ConnectedThings
        {
            get { return new List<Thing>(Connections.Values); }
        }

	    public int ConnectedThingsCount
	    {
			get { return Connections.Count; }
	    }


        /**
         * Compare if two things are equals
         * 
         * Can compare also connected things, without infinite recursions.
         */
        public bool Compare(Thing other, bool compareId = true, bool deepComparaisonForConnectedThings = false)
        {
            // Optimization, when two things are the same
            if (this == other)
            {
                return true;
            }

            // If the types are not the sames
            // The type can be null so it's a bit ugly here
            if (other == null || (Type != null && other.Type != null && Type.Name != other.Type.Name) ||
                (Type == null && other.Type != null) || (Type != null && other.Type == null))
            {
                return false;
            }

            // If we just need to compare the ids, it's very fast
            if (compareId && _id != other._id)
            {
                return false;
            }

            // Check if the connection list are the same
            if (Connections.Count != other.Connections.Count ||
                Connections.Keys.Any(key => !other.Connections.ContainsKey(key)))
            {
                return false;
            }

            if (Properties.Count != other.Properties.Count)
            {
                return false;
            }

            // Check if the properties are the same
            foreach (var property in Properties)
            {
                Property otherProp;
                other.Properties.TryGetValue(property.Key, out otherProp);

                if (otherProp == null || !property.Value.CompareValue(otherProp))
                {
                    return false;
                }
            }

            // And start the start recursion if needed
            if (deepComparaisonForConnectedThings)
            {
                // The hashset is declared only once and here
                return RecursiveCompare(other, new HashSet<string>());
            }

            return true;
        }

        // Do a recursive comparaison, should be called 
        private bool RecursiveCompare(Thing other, HashSet<string> workIds) {
            // If the thing was already checked, we don't need to check it again
            if (workIds.Contains(_id))
            {
                // And it's alway true, because we are still looking for a difference
                return true;
            }

            // Made a simple comparison first
            if (!Compare(other))
            {
                return false;
            }

            // We are working on the current thing,
            // this instruction prevent infinite recursion
            workIds.Add(_id);

            // And start the recursion for connected things
            foreach (var connectedThing in Connections)
            {
                var otherThing = other.Connections[connectedThing.Key];

                if (!connectedThing.Value.RecursiveCompare(otherThing, workIds))
                {
                    return false;
                }
            }

            return true;

        }

        public IEnumerable<Property> GetProperties()
        {
            // Create a copy of this list, so the user can change the
            // things properties during a loop on this list without exceptions
            return new List<Property>(Properties.Values);
        }


	    private BuildANewThing.ThingPropertyBuilder _propertyBuilder;
	    public BuildANewThing.ThingPropertyBuilder ContainingA
	    {
		    get { return _propertyBuilder ?? (_propertyBuilder = new BuildANewThing.ThingPropertyBuilder(this)); }
	    }
	    
		public BuildANewThing.ThingPropertyBuilder ContainingAn
	    {
		    get { return _propertyBuilder ?? (_propertyBuilder = new BuildANewThing.ThingPropertyBuilder(this)); }
	    }

	    public String String(String key)
	    {
		    var p = GetProperty<Property.String>(key);
		    return p != null ? p.Value : string.Empty;
	    }

	    public Thing String(String key, String value)
	    {
		    SetProperty(new Property.String(key, value));
		    return this;
	    }

	    public double Double(String key)
	    {
		    var p = GetProperty<Property.Double>(key);
		    return p != null ? p.Value : default(double);
	    }
	    
		public Thing Double(String key, double value)
	    {
		    SetProperty(new Property.Double(key, value));
		    return this;
	    }
	    
		public int Int(String key)
	    {
		    var p = GetProperty<Property.Int>(key);
		    return p != null ? p.Value : default(int);
	    }
	    
		public Thing Int(String key, int value)
	    {
		    SetProperty(new Property.Int(key, value));
		    return this;
	    }
		
		public bool Boolean(String key)
	    {
		    var p = GetProperty<Property.Boolean>(key);
		    return p != null ? p.Value : default(bool);
	    }
	    
		public Thing Boolean(String key, bool value)
	    {
		    SetProperty(new Property.Boolean(key, value));
		    return this;
	    }
		
		public DateTime DateTime(String key)
	    {
		    var p = GetProperty<Property.DateTime>(key);
			return p != null ? p.Value : new DateTime(0);
	    }
	    
		public Thing DateTime(String key, DateTime value)
	    {
		    SetProperty(new Property.DateTime(key, value));
		    return this;
	    }

	    public Location.Point LocationPoint(String key = "location")
	    {
		    var p = GetProperty<Property.Location.Point>(key);
			return p != null ? p.Value : new Location.Point();
	    }

	    public Thing LocationPoint(String key, Location.Point value)
	    {
		    SetProperty(new Property.Location.Point(key, value));
		    return this;
	    }
	    
		public Location.LatLng LocationLatLng(String key = "location")
	    {
		    var p = GetProperty<Property.Location.LatLng>(key);
			return p != null ? p.Value : new Location.LatLng();
	    }

	    public Thing LocationLatLng(String key, Location.LatLng value)
	    {
		    SetProperty(new Property.Location.LatLng(key, value));
		    return this;
	    }
		
		public Location.Equatorial LocationEquatorial(String key = "location")
	    {
		    var p = GetProperty<Property.Location.Equatorial>(key);
			return p != null ? p.Value : new Location.Equatorial();
	    }

	    public Thing LocationEquatorial(String key, Location.Equatorial value)
	    {
		    SetProperty(new Property.Location.Equatorial(key, value));
		    return this;
	    }
    }
}

File: TypeScript/Thing.ts
/// <reference path="./bower_components/DefinitelyTyped/underscore/underscore.d.ts" />

module ThingModel {
	export class Thing {

		/**
		 *	String ID, unique in the world.
		 *
		 *	It can't be null.
		 *	Two things objets with the same ID represent the same thing.
		 */
		private _id : string;

		public get ID() : string {
			return this._id;
		}


		/**
		 *	Reference to the object type.
		 *
		 *	The type is null by default, and the default type is applied.
		 */
		private _type : ThingType = null;

		public get Type() : ThingType {
			return this._type;
		}

		/**
		 *	A thing contains a list of properties.
		 *
		 *	A property is defined by his key and his value.
		 */
		private _properties: { [key: string]: Property };
		private _propertiesCount: number;

		/**
		 *	A thing can have a list of connected things.
		 *
		 *	But a thing cannot be connected directly to itself.
		 */
		private _connections: { [key: string]: Thing };
		private _connectionsCount: number;

		constructor(id : string, type : ThingType = null) {
			if (!id) {
				throw new Error("The thing ID should not be null or empty");
			}

			this._id = id;
			this._type = type;

			this._properties = {};
			this._propertiesCount = 0;
			this._connections = {};
			this._connectionsCount = 0;
		}

		public SetProperty(property: Property): void {
			if (!this.HasProperty(property.Key)) {
				++this._propertiesCount;
			}
			this._properties[property.Key] = property;
		}

		public HasProperty(key : string) : boolean {
			return _.has(this._properties, key);
		}

		public GetProperty<T extends Property>(key: string, type?: Type): T {

			var prop = this._properties[key];
			if (!prop || (type && prop.Type != type)) {
				return null;
			}

			return <any>prop;
		}

		public GetString(key: string) {

			var prop = this.GetProperty<ThingModel.Property.String>(key, Type.String);

			if (!prop) {
				return null;
			}

			return prop.Value;
		}

		public Connect(thing: Thing): void {
			if (!thing) return;

			if (thing === this || thing._id === this._id) {
				throw new Error("You can't connect a thing directly to itself");
			}
			if (!this.IsConnectedTo(thing)) {
				++this._connectionsCount;
			}
			this._connections[thing._id] = thing;
		}

		public Disconnect(thing: Thing): boolean {
			if (this.IsConnectedTo(thing)) {
				--this._connectionsCount;	
				delete this._connections[thing._id];
				return true;
			}
			return false;
		}

		public DisconnectAll() : void {
			this._connections = {};
			this._connectionsCount = 0;
		}

		public IsConnectedTo(thing: Thing): boolean {
			// !!thing in order will convert the assertion to a boolean
			return !!thing && _.has(this._connections, thing._id);
		}

		public get ConnectedThings(): Thing[] {
			return _.values(this._connections);
		}

		public get ConnectedThingsCount(): number {
			return this._connectionsCount;
		}

		public get Properties(): Property[] {
			return _.values(this._properties);
		}

		public Compare(other: Thing, compareId: boolean = true,
			deepComparisonForConnectedThings: boolean = false): boolean {
			// Optimization, when two things are the same instance
			if (this === other) {
				return true;
			}

			// If the types are not the same
			if (!other || (this._type != null && other._type != null &&
				this._type.Name != other._type.Name) ||
				(this._type == null && other._type != null) ||
				(this._type != null && other._type == null)) {
				return false;
			}

			// If we need to compare the ids, and they are not the same
			if (compareId && this._id != other._id) {
				return false;
			}

			// Check if the connections are the same
			if (this._connectionsCount !== other._connectionsCount || 
				_.any(this._connections, (connectedThing:Thing)=> {
					return !_.has(other._connections, connectedThing._id);
				})) {
				return false;
			}

			// Check if the properties are the sames
			if (this._propertiesCount !== other._propertiesCount ||
				_.any(this._properties, (property: Property) => {
					var otherProp = other._properties[property.Key];

					return otherProp == null || !otherProp.CompareValue(property);
				})) {
				return false;
			}

			if (deepComparisonForConnectedThings) {
				return this.RecursiveCompare(other, {});
			}

			return true;
		}

		private RecursiveCompare(other: Thing,
			alreadyVisitedObjets: { [id: string]: boolean }): boolean {

			// If the thing was already checked,
			// we don't need to check it again
			if (_.has(alreadyVisitedObjets, this._id)) {

				// It's true because we are still looking for a difference
				return true;
			}

			// Made a simple comparison first
			if (!this.Compare(other)) {
				return false;
			}

			// Register the thing now, prevent infinite recursions
			alreadyVisitedObjets[this._id] = true;

			return !_.any(this._connections, (connectedThing: Thing)=> {
				var otherThing = other._connections[connectedThing._id];

				return !connectedThing.RecursiveCompare(otherThing, alreadyVisitedObjets);
			});
		}

		private _propertyBuilder: ThingPropertyBuilder;
		public get ContainingA(): ThingPropertyBuilder {
			if (!this._propertyBuilder) {
				this._propertyBuilder = new ThingPropertyBuilder(this);
			}
			return this._propertyBuilder;
		}

		public get ContainingAn(): ThingPropertyBuilder {
			return this.ContainingA;
		}

		public String(key: string): string;
		public String(key: string, value: string): Thing;
		public String(key: string, value?: string): any {
			if (value) {
				this.SetProperty(new Property.String(key, value));
				return this;
			} else {
				var p = this.GetProperty(key, Type.String);
				if (!p) {
					return null;
				}
				return (<Property.String>p).Value;
			}
		}

		public Double(key: string): number;
		public Double(key: string, value: number): Thing;
		public Double(key: string, value?: number): any {
			if (value) {
				this.SetProperty(new Property.Double(key, value));
				return this;
			} else {
				var p = this.GetProperty(key, Type.Double);
				if (!p) {
					return null;
				}
				return (<Property.Double>p).Value;
			}
		}

		public Int(key: string): number;
		public Int(key: string, value: number): Thing;
		public Int(key: string, value?: number): any {
			if (value) {
				this.SetProperty(new Property.Int(key, value));
				return this;
			} else {
				var p = this.GetProperty(key, Type.Int);
				if (!p) {
					return null;
				}
				return (<Property.Int>p).Value;
			}
		}

		public Boolean(key: string): boolean;
		public Boolean(key: string, value: boolean): Thing;
		public Boolean(key: string, value?: boolean): any {
			if (value) {
				this.SetProperty(new Property.Boolean(key, value));
				return this;
			} else {
				var p = this.GetProperty(key, Type.Boolean);
				if (!p) {
					return null;
				}
				return (<Property.Boolean>p).Value;
			}
		}

		public DateTime(key: string): Date;
		public DateTime(key: string, value: Date): Thing;
		public DateTime(key: string, value?: Date): any {
			if (value) {
				this.SetProperty(new Property.DateTime(key, value));
				return this;
			} else {
				var p = this.GetProperty(key, Type.DateTime);
				if (!p) {
					return null;
				}
				return (<Property.DateTime>p).Value;
			}
		}

		public LocationPoint(): Location.Point;
		public LocationPoint(key: string): Location.Point;
		public LocationPoint(value: Location.Point): Thing;
		public LocationPoint(key: string, value: Location.Point): Thing;

		public LocationPoint(mixed?: any, value?: Location.Point): any {
			var key: string;

			if (typeof mixed === 'string' || mixed instanceof String) {
				key = mixed;
			} else {
				key = "location";
				value = mixed;
			}

			if (value) {	
				this.SetProperty(new Property.Location.Point(key, value));
				return this;
			} else {
				var p = this.GetProperty(key, Type.LocationPoint);
				if (!p) {
					return null;
				}
				return (<Property.Location.Point>p).Value;
			}
		}

		public LocationLatLng(): Location.LatLng;
		public LocationLatLng(key: string): Location.LatLng;
		public LocationLatLng(value: Location.LatLng): Thing;
		public LocationLatLng(key: string, value: Location.LatLng): Thing;

		public LocationLatLng(mixed?: any, value?: Location.LatLng): any {
			var key: string;

			if (typeof mixed === 'string' || mixed instanceof String) {
				key = mixed;
			} else {
				key = "location";
				value = mixed;
			}

			if (value) {	
				this.SetProperty(new Property.Location.LatLng(key, value));
				return this;
			} else {
				var p = this.GetProperty(key, Type.LocationLatLng);
				if (!p) {
					return null;
				}
				return (<Property.Location.LatLng>p).Value;
			}
		}

		public LocationEquatorial(): Location.Equatorial;
		public LocationEquatorial(key: string): Location.Equatorial;
		public LocationEquatorial(value: Location.Equatorial): Thing;
		public LocationEquatorial(key: string, value: Location.Equatorial): Thing;

		public LocationEquatorial(mixed?: any, value?: Location.Equatorial): any {
			var key: string;

			if (typeof mixed === 'string' || mixed instanceof String) {
				key = mixed;
			} else {
				key = "location";
				value = mixed;
			}

			if (value) {	
				this.SetProperty(new Property.Location.Equatorial(key, value));
				return this;
			} else {
				var p = this.GetProperty(key, Type.LocationEquatorial);
				if (!p) {
					return null;
				}
				return (<Property.Location.Equatorial>p).Value;
			}
		}
	}	
}